JoinColony/colonyCDapp
# [Arbitrary txs] Improve function data decoding

## Description

Don't be put off by the number of files changed, 80% of them are part of a small function arguments refactor! 

See #3875 for context.

This PR shifts the function data decoding to the UI and makes it rely on the original ABI used to create the arbitrary transaction.

[block-ingestor PR](https://github.com/JoinColony/block-ingestor/pull/305)

## Testing

To test the decoding without hacking a sandbox to call the util, we'll deploy a test contract that has functions with a variety of argument types we expect to encounter.

1. Ensure your dev env is fully running.
2. Create a new file called `TestContract.sol` in your CDapp root directory with the following contents:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract TestContract {
    function simpleFunction(uint256 uint256Arg) public {}

    function anotherSimpleFunction(
        string memory stringArg,
        address addressArg
    ) public {}

    function testFunction(
        address addressArg,
        string memory stringArg,
        bool boolArg,
        uint8 uint8Arg,
        uint256 uint256Arg,
        bytes memory bytesArg
    ) public {}
}
```
4. Run the following in your terminal to copy the contract over to the network container and compile it:
```bash
docker cp TestContract.sol network:/colonyCDappBackend/colonyNetwork/contracts && docker exec -it -w /colonyCDappBackend/colonyNetwork network npx hardhat compile
```
5. Start the hardhat console with `npm run hardhat`, then run the following code line-by-line:
```js
cf = await ethers.getContractFactory('TestContract')
c = await cf.deploy()
c.address
```
The last expression will print the address of the newly deployed contract. Copy it for later use.

In the UI:
1. Create a custom transaction action (under Manage colony).
2. In the Add transaction modal, paste the address of the deployed contract.
3. Because there's no API to get ABI for local contracts, paste it in the textarea manually:
```json
[
  {
    "inputs": [
      {
        "internalType": "string",
        "name": "stringArg",
        "type": "string"
      },
      {
        "internalType": "address",
        "name": "addressArg",
        "type": "address"
      }
    ],
    "name": "anotherSimpleFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "uint256Arg",
        "type": "uint256"
      }
    ],
    "name": "simpleFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "addressArg",
        "type": "address"
      },
      {
        "internalType": "string",
        "name": "stringArg",
        "type": "string"
      },
      {
        "internalType": "bool",
        "name": "boolArg",
        "type": "bool"
      },
      {
        "internalType": "uint8",
        "name": "uint8Arg",
        "type": "uint8"
      },
      {
        "internalType": "uint256",
        "name": "uint256Arg",
        "type": "uint256"
      },
      {
        "internalType": "bytes",
        "name": "bytesArg",
        "type": "bytes"
      }
    ],
    "name": "testFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  }
]
```
4. Select one of the available functions and fill its arguments with data of your choice (the validation should help you out!).
6. Add more transactions or create an action with just one.

You will be taken to a rudimentary UI showing the details of the arbitrary transactions your action contained:
- method name
- argument names with their types 
- values of the arguments

<img width="1027" alt="image" src="https://github.com/user-attachments/assets/627dc5b5-d4aa-43d0-9db4-f040ca6ea0bf" />

Resolves #3875
